### PR TITLE
LP-1612: Add an api to fetch pre and post assessment score

### PR DESCRIPTION
**Pull request by zeshanarifios**
[LP-1612](https://philanthropyu.atlassian.net/browse/LP-1612)

### CODE REVIEW

File: lms/urls.py [hunk 1]
```diff
@@ -1012,6 +1013,15 @@
         include('lms.djangoapps.certificates.apis.urls', namespace='certificates_api')),
 ]
 
+# Fetch assessments score api
+urlpatterns += [
+    url(r'^api/get_assessments_score/{}/(?P<chapter_id>[^/]*)/'.format(
```
Comment (zahid-arbisoft) [inline lms/urls.py]: If possible, we should remove get from url, HTTP GET request is enough to indicate that

File: openedx/features/courseware/constants.py [hunk 2]
```diff
@@ -0,0 +1,5 @@
+"""Constants for courseware"""
+
+
```
Comment (zahid-arbisoft) [inline openedx/features/courseware/constants.py]: Lets remove extra empty line, see [reference](https://github.com/django/django/blob/master/django/db/models/constants.py) file from django as sample

Comment (zeshanarifios) [reply openedx/features/courseware/constants.py]: Removed
@zahid-arbisoft

File: lms/urls.py (hunk 1 above)
Comment (zeshanarifios) [reply lms/urls.py]: Removed
@zahid-arbisoft

File: openedx/features/courseware/views.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    pre_assessment_score = post_assessment_score = None
+    chapter_found = pre_assessment_attempted = False
+
+    for chapter in courseware_summary:
+        if chapter['url_name'] == chapter_id:
+            chapter_found = True
+            for section in chapter['sections']:
+                section_format = (section.format.lower()
+                                  if section.format is not None
```
Comment (ahmed-faraz46) [inline openedx/features/courseware/views.py]: I guess we can change this to just if `section.format else`

File: openedx/features/courseware/views.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+from lms.djangoapps.courseware.exceptions import CourseAccessRedirect
+from lms.djangoapps.courseware.courseware_access_exception import \
+                                                CoursewareAccessException
+from lms.djangoapps.grades.course_grade_factory import CourseGradeFactory
+from .constants import PRE_ASSESSMENT_FORMAT, POST_ASSESSMENT_FORMAT
+
+
+@api_view()
+@authentication_classes([SessionAuthentication])
+@permission_classes([IsAuthenticated])
```
Comment (saad-waqar) [inline openedx/features/courseware/views.py]: ```@view_auth_classes(is_authenticated=True)```
I found this in /edx-platform/lms/djangoapps/courseware/module_render.py
I'm not sure if this is directly applicable here can you check?

File: openedx/features/courseware/views.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+from opaque_keys.edx.keys import CourseKey
+
+from lms.djangoapps.courseware.exceptions import CourseAccessRedirect
+from lms.djangoapps.courseware.courseware_access_exception import \
+                                                CoursewareAccessException
+from lms.djangoapps.grades.course_grade_factory import CourseGradeFactory
+from .constants import PRE_ASSESSMENT_FORMAT, POST_ASSESSMENT_FORMAT
+
+
+@api_view()
```
Comment (saad-waqar) [inline openedx/features/courseware/views.py]: ```@api_view(['GET'])```

File: openedx/features/courseware/views.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    courseware_summary = course_grade.chapter_grades.values()
+
+    pre_assessment_score = post_assessment_score = None
+    chapter_found = pre_assessment_attempted = False
+
+    for chapter in courseware_summary:
+        if chapter['url_name'] == chapter_id:
+            chapter_found = True
+            for section in chapter['sections']:
+                section_format = (section.format.lower()
```
Comment (saad-waqar) [inline openedx/features/courseware/views.py]:
```
if not section.format:
    continue
```

File: openedx/features/courseware/views.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zeshanarifios) [reply openedx/features/courseware/views.py]: I think it isn't needed, 
By default only GET methods will be accepted. Other methods will respond with "405 Method Not Allowed". 
Reference: https://www.django-rest-framework.org/api-guide/views/#api_view

File: openedx/features/courseware/views.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    except (CourseAccessRedirect, CoursewareAccessException):
+        return Response({
+                'detail': _('User does not have access to this course')
+                }, status=status.HTTP_403_FORBIDDEN)
+
+    course_grade = CourseGradeFactory().read(request.user, course)
+    courseware_summary = course_grade.chapter_grades.values()
+
+    pre_assessment_score = post_assessment_score = None
+    chapter_found = pre_assessment_attempted = False
```
Comment (saad-waqar) [inline openedx/features/courseware/views.py]:
```try:
    chapter = next(obj for chapter in courseware_summary if chapter['url_name'] == chapter_id)
except StopIteration:
    return Response({'detail': _('Chapter id is not valid')}, status=status.HTTP_400_BAD_REQUEST)

Comment (saad-waqar) [reply openedx/features/courseware/views.py]: adjust line length accordingly I cant do that here

File: openedx/features/courseware/views.py (hunk 5 above)
Comment (zeshanarifios) [reply openedx/features/courseware/views.py]: I think it isn't needed,
By default only GET methods will be accepted. Other methods will respond with "405 Method Not Allowed".
Reference: https://www.django-rest-framework.org/api-guide/views/#api_view

File: openedx/features/courseware/views.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        if chapter['url_name'] == chapter_id:
+            chapter_found = True
+            for section in chapter['sections']:
+                section_format = (section.format.lower()
+                                  if section.format is not None
+                                  else '')
+                if section_format == PRE_ASSESSMENT_FORMAT:
+                    pre_assessment_score = section.all_total.earned
+                    pre_assessment_attempted = \
+                        section.all_total.first_attempted is not None
```
Comment (saad-waqar) [inline openedx/features/courseware/views.py]: bool(section.all_total.first_attempted)

File: openedx/features/courseware/views.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+from courseware.courses import get_course_with_access
+from opaque_keys.edx.keys import CourseKey
+
+from lms.djangoapps.courseware.exceptions import CourseAccessRedirect
+from lms.djangoapps.courseware.courseware_access_exception import \
+                                                CoursewareAccessException
+from lms.djangoapps.grades.course_grade_factory import CourseGradeFactory
+from .constants import PRE_ASSESSMENT_FORMAT, POST_ASSESSMENT_FORMAT
+
```
Comment (mabdullahz) [inline openedx/features/courseware/views.py]: I think there should be a line break before the `.constants` import

**Pull request merged**